text/0000-dockerfiles.md

```diff
@@ -0,0 +1,224 @@
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with dynamically-generated build-time and runtime Dockerfiles that act as pre-build base image extensions.
+These extensions would participate in detection and execute before the buildpack build process.
+
+For a given application, a build that uses extensions could be optimized by creating a more narrowly-scoped builder that does not contain extensions.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. The order of operations would be something like the following:
+* `analyze`
+* `detect` - after standard detection detect will also run extensions' bin/generate, output Dockerfiles are written to a volume
+* `extend` - applies run.Dockerfiles to run image (could run in parallel with builder image extension)
+* `extend` - applies build.Dockerfiles to builder image
+* `restore`
+* `build`
+* `export`
+
+When Dockerfiles are used to update the run image, care should be taken to ensure registry access prior to the `build` phase, to avoid long builds that fail due to incorrect credentials.
+
+### Dynamically-applied Dockerfiles
+
+A builder image may include any number of "extensions" directories in `/cnb/extensions/`.
+
+Extensions are similar to buildpacks: they have two executables: `/bin/detect` and `/bin/generate`. The interface for these executables is similar to a buildpack's `/bin/detect` and `/bin/build`.
+However, instead of a `buildpack.toml` file, extensions have a `extension.toml` file:
+```toml
+api = "<buildpack API version>"
+
+[extension]
+id = "<extension ID>"
+name = "<extension name>"
+version = "<extension version>"
+homepage = "<extension homepage>"
+description = "<extension description>"
+keywords = [ "<string>" ]
+
+[[extension.licenses]]
+type = "<string>"
+uri = "<uri>"
+```
+
+Extensions may be packaged and examined similar to buildpacks, but with analogous `pack extension` subcommands.
+
+Other pack CLI commands, such as `pack builder create`, will be extended to include support for extensions.
+
+Unlike buildpacks,
+- Extensions must not be included in a meta-buildpacks
+- Extensions must not have `order`/`group` definitions in `extension.toml`
+
+Extensions participate in the buildpack detection process, with the same UID, GID, and interface for `/bin/detect`.
+However,
+- `/bin/detect` is optional for extensions, and they are assumed to pass detection when it is not present. Just like with buildpacks, a /bin/detect that exits with a 0 exit code passes detection, and fails otherwise.
+- If an extension is missing `/bin/detect`, the extension root `./detect` directory is treated as a pre-populated output directory (i.e., extensions can include a static build plan).
```

Comment:
```suggestion
Extensions participate in the buildpack detection process, with the same UID, GID, and interface for `./bin/detect`.
However,
- `./bin/detect` is optional for extensions, and they are assumed to pass detection when it is not present. Just like with buildpacks, a `./bin/detect` that exits with a 0 exit code passes detection, and fails otherwise.
- If an extension is missing `./bin/detect`, the extension root `./detect` directory is treated as a pre-populated output directory (i.e., extensions can include a static build plan).
```